ls1intum/Artemis
# `Exam mode`: Allow editing individual working time during conduction

### Checklist
#### General


#### Server
#### Client
#### Changes affecting Programming Exercises

### Motivation and Context


Fix #5962

### Description

It is now possible to change the working time of individual student exams during their conduction. The change is also sent via web socket to the student and displayed as an alert at the top right of the screen.
Programming exercises should get locked at the new due date for the student.

### Steps for Testing

Prerequisites:
- 1 Instructor
- 2 Students
- 1 Exam with a programming exercise.

1. Log in to Artemis
2. Create the exam
3. Participate as both students (you can do this simultaneously)
4. Go to the student exam overview as instructor
5. Decrease the working time of one student and verify that the student gets a notification (e.g. use two browsers one for the student, one for the instructor)
6. Increase the working time of the other student and verify that the student gets a notification (e.g. use two browsers one for the student, one for the instructor)
7. Make surre that both repositories are locked at their respective due date and the student can no longer push

### Review Progress






#### Code Review
#### Manual Tests

### Test Coverage


| Class/File | Branch | Line |
|------------|-------:|-----:|
| ProgrammingExerciseScheduleService.java | 75% | 87% |
| exam-participation.component.ts | 67% | 78% |

### Screenshots


Instructor view:
<img width="516" alt="Bildschirm­foto 2023-01-28 um 01 08 38" src="https://user-images.githubusercontent.com/38322605/215228594-2f145b0b-cf82-4cd7-b7c0-0d2b187a528a.png">
Student view when time was changed:
<img width="1680" alt="Bildschirm­foto 2023-01-28 um 01 09 36" src="https://user-images.githubusercontent.com/38322605/215228623-c5edda0b-9f63-4714-be51-64f783d01ed1.png">
